Documentation: Re-write setup of development environment sections

groundhogxc opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Documentation**
  * Renamed Linux-centric guide to cross-platform "Setting up a development environment" and broadened scope to Android, KOBO, UNIX, Windows, etc.
  * Reorganized into top-level sections for Linux, Windows (WSL2), Vagrant, Docker, and detailed setup guides.
  * Added explicit Windows/WSL2 instructions, Vagrant VM and Docker container build paths, and editor/LaTeX guidance.
  * Clarified VM lifecycle, container caveats, X11/remote editing, Git/line-ending and provisioning notes.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

groundhogxc commented:
Tried to improve the setup section; any input/comments is welcome! (I can work on it more if needed before, or after, merging)
Also, if someone with a Mac could write a short Mac description, could be nice.

groundhogxc commented:
> There is another methode to build xcsoar which i find very easy and might be the simplest for unexperienced users.
> 
> Namly using github workflows, you can have xcsoar build on github when you for example push a certain branch. Much less trouble than the VM methode.

Sounds very interesting! Maybe write a short description of how to set that up? (Ideally including how to get the compile logs back into the IDE) 
BTW, does this require some sort of github subscription or so? Someone has to pay for the computing power, I guess?

groundhogxc commented:
Sounds pretty good! Following up on errors might be a bit cumbersome, and, of course, actually running the (UNIX) program still requires the VM on the computer, but that could be a simpler version of the VM potentially.

lordfolken commented:
Thanks for taking the time. Ill review it.

lordfolken commented:
I'm grateful for the effort, but I really don't like it. 

1. The text is in places quite patronizing. (If you don't know what you are doing. Its really up to the user to decide when they are in over their heads)
2. This is supposed to be documentation and not prose text. So short and concise and to the point. 
3.  The primary development system is Debian linux and i think we should stick to that supported path. 
4. The vagrant and docker images are convenience to have a clean room compile/debug environment that is reproducable.
5.  We shouldn't introduce people to git and how it works, there is plenty of howtos on those topics. 



On the up side: thanks for documenting docker.

groundhogxc commented:
Thanks for the feedback!

>     1. The text is in places quite patronizing. (If you don't know what you are doing. Its really up to the user to decide when they are in over their heads)
No problem, I can work on that.
 
>     2. This is supposed to be documentation and not prose text. So short and concise and to the point.
OK, I can try and compress it.

>     3. The primary development system is Debian linux and i think we should stick to that supported path.
Hm, are you referring to Windows as the host system?

>     4. The vagrant and docker images are convenience to have a clean room compile/debug environment that is reproducable.
Yes. Did I accidentally say something else?

>     5. We shouldn't introduce people to git and how it works, there is plenty of howtos on those topics.
I thought about that when I did that; same with how much basic intro to docker there should be.

My opinion here is that there should be enough very concrete instructions to make things work for a user who does not know how it works. Basically the idea is that even a non-programmer can get far enough to contribute to e.g. the manuals in a reasonably efficient way. Because this makes a lot more potential manpower available. When I started working on code and manuals, I was in this situation myself - I basically learned git from scratch in order to work with XCSoar files. So while I absolutely agree there should not be a serious howto on git, I think the steps described should cover enough to help someone who is interested in contributing, but not really interested in the tools themselves. After all, the programming experts anyway will not need this entire section (they know the conventions or will just look into the Vagrantfile itself), this is all abut making it easier for people with relevant know-how and time to contribute, but not experienced in complex programming projects.
At least those are my thoughts about this...

lordfolken commented:
> > ```
> > 3. The primary development system is Debian linux and i think we should stick to that supported path.
> > ```
> Hm, are you referring to Windows as the host system?
I want to say that if you want to develop for xcsoar you should start with a Debian Linux 12, use the provisioning scripts and build the UNIX target first, and work / develop with that.

This can be native, or in a VM.

The docker container is designed to support as many targets as possible, and is kind of a sanity check (offline ci-pipline equivalent), or if someone comes from a different distribution. Or also if you don´t want to pollute your local system with GBs of dependencies. 



> > ```
> > 4. The vagrant and docker images are convenience to have a clean room compile/debug environment that is reproducable.
> > ```
> 
> Yes. Did I accidentally say something else?
No its fine.

 
> > ```
> > 5. We shouldn't introduce people to git and how it works, there is plenty of howtos on those topics.
> > ```
> 
> I thought about that when I did that; same with how much basic intro to docker there should be.
> 

> My opinion here is that there should be enough very concrete instructions to make things work for a user who does not know how it works. Basically the idea is that even a non-programmer can get far enough to contribute to e.g. the manuals in a reasonably efficient way. 

I'm game, lets try it. 

Maybe make a couple of commits, then we can merge the ones we agree on, and discuss the rest.

DanD222 commented:
My 0.02 relating to git - I think it's also important to elaborate a bit more on what a Pull Request should look like during its lifecycle, with specific focus on rebase - i.e. there being no "addressing changes" or "Merge master into feature branch" commits in the PR. It'd also be good to mention Force Push so that people don't close a PR and start a new one when a change is requested.

I feel a lot of people try to use the web interface for submitting patches, which is fine until a change request comes.

If you feel like writing that up then by all means, if you'd like to split the load I could probably have a go at something.

groundhogxc commented:
> My 0.02 relating to git - I think it's also important to elaborate a bit more on what a Pull Request should look like during its lifecycle, with specific focus on rebase - i.e. there being no "addressing changes" or "Merge master into feature branch" commits in the PR. It'd also be good to mention Force Push so that people don't close a PR and start a new one when a change is requested.

Absolutely! I think the section which is more related to that is the "Policy" section. Some of what you mention I think is already there. You could have a look what is missing (in fact some stuff there seems to bee outdated - like the two master branch structure, which was given up some time after V7)

With folken's comment that things should be short and snappy - this could also be done in a way where the policy is described concisely there, and then there is some "howto" or "introduction" section somewhere further in the back.

groundhogxc commented:
Alright, sorry for the long gap, I was swamped in work - but more importantly I actually changed my own recommendation with respect to how to work on Windows totally :-)

I now recommend to use WSL on Windows directly (no docker) as the standard method for Windows users (who do not want to deal with Gnome or KDE or so). WSL performs FAR better than Virtualbox etc. at the moment, and is not as hard to use as it used to be. However, I still need to complete the WSL section :-)

So this is still work in progress. But I want to still upload it here to collect feedback again: This is now very differently structured compared to before, and many sections have also been significantly rewritten.

The new structure is that the main part is actually quite short, with the four sections "Build setup on Linux", "Build setup on Windows", "how to use the Vagrantfile provided" and "How to use our containers". So it is centered around the things that is provided from us (provisioning scripts, Vagrantfile, dockerfile), together with the two Main OSes.

This is then followed by the longer step-by-step instructions aimed more towards people who know how to code but find ethe build setup stuff intimidationg (like myself and other people I know). This is devided into a Linux section and a windows section.

Basically what I am looking for is some feedback on this structure, and whether the style is better (in particular WRT @lordfolken 's suggestions) - and of course for any other kind of feedback/corrections you want to make before I put in the work to finalize the PR.

groundhogxc commented:
Ah, and I forgot: It would be great i f someone with an actual deep understanding of the toolchains could check if the statements are correct for docker:

- there is an existing container ready for download on ghcr.io, which is the easiest way to do it. This is used by all the commands for building / running described in the ide/docker/README.md file. This does not need a dockerfile.
- Separately, there is the Dockerfile, to build *this exact same* container locally from scratch, starting with a pure Debian. This is effectively what the section with the "docker build" command the README describes.
- If I wanted to go back to an earlier version of the code, e.g. for bisecting a problem, I might have to rebuild the container locally in order to get to the earlier versions of libraries, Android tools, etc., if/when that is required. Otherwise, the ghcr.io one is always "latest". Correct?

(By the way, in this respect: I think I saw also other places in the codebase sometimes uses some kind of "latest" formulation in including tools or libraries. This might reconstructing an older version harder, as opposed to always including/provisioning explicit versions of tools and libraries everywhere)

lordfolken commented:
@coderabbitai fullreview

groundhogxc commented:
Hi - sorry, I did not log in into github for a long time, and completely missed the feedback you sent a couple months ago. Thanks!

I went through the whole thing and made the suggested changes from the first two coderabbit posts, I think. Also, I reviewed it thoroughly another time for formulation, typos, consistency and so on.

Coderabbit also flagged the two questions that I also posed earlier because I am not 100% sure myself:

   * Is the ghcr.io container indeed the "easiest way" and "most exact replication"?
   * Are the relationships between ghcr.io container, local Dockerfile, and version compatibility accurately described?

I think it is correct the way it is written (in the sense that the Docker build with those images is the most faithful build environment to the official build, because that is used in the CI builds)? But if not, let me know.

I also implemented the suggestions by coderabbit "diffs", but mostly the last two. The others seemed a bit off (except for the obvious typos of course). Also, the last two needed significant fixing/modification.

So far I did not do any reorganization of the commits, I guess in the end it should be squashed into one or two commits? (and rebased again of course)

groundhogxc committed:
Documentation: Devsetup: some updates for 2025

groundhogxc committed:
Documentation: Re-Structered the develop system setup Manual, improved most versions.
The new structure is now a main part with four short blocks (Build system on Linux using provisioning scripts, Build on WIndows using WSL2, how to use the Vagrantfile provided, and how to use the Container provided by XCsoar). This give the main information, and should be enough for experienced users. This is followed by step-by-step instructions for how to get started, intended for less experienced users. This section has two main parts Linux / Windows.

groundhogxc committed:
Documentation: Dev Setup - review corrections for typos and other smaller errors

groundhogxc committed:
Documentation: Devsetup: small fixes and typos, from coderabbit and manual review

groundhogxc committed:
Documentation: Devsetup - implement fixes

groundhogxc committed:
Documentation: Re-write setup of development environment sections
gives detailed instructions especially for using Vagrant, discusses both Linux and Windows explicitely

groundhogxc committed:
Documentation: Devsetup: Documentation: Devsetup: some updates for 2026

lordfolken commented:
**Review (doc only)**

Thanks for the substantial update to `doc/devsetup.rst` — the extra coverage for WSL2, Vagrant, and Docker matches how people actually set things up, and dropping the ancient Eclipse version notes is a clear win.

A few things I’d like to see before merge:

1. **Structure / audience** — Please make the “happy path” unmistakable up front: Debian-style Linux → `ide/provisioning` → build `TARGET=UNIX` first (native or VM), then present WSL2 / Vagrant / Docker as alternatives. That matches the direction from the earlier discussion about keeping the doc reference-style and Debian-centric.

2. **Length and tone** — Some sections read quite tutorial-like (especially Windows + git setup). Tightening to short, imperative steps where possible would help readers who already know git, without losing the concrete commands newcomers need.

3. **Small fixes** — Typo: “componenents” → “components”. Consider standardizing on one clone flag (`--recursive` vs `--recurse-submodules`). In “Which method should I use?”, clarify the Vagrant case so it doesn’t sound like the build runs on the Windows filesystem when the VM flow uses a guest working tree — wording should match the Vagrant section above.

4. **RST** — After edits, a quick pass with a built manual (`make manual` or whatever CI uses) to ensure literal blocks and headings still render as intended.

Optional nice follow-up (separate or small subsection): a brief note on using GitHub Actions for builds, as mentioned in the thread — not a replacement for a local dev env, but useful for some contributors.

Happy to re-review after a revision.
